### PR TITLE
Mongo connector topn pushdown

### PR DESCRIPTION
**Pull request by NikhilDesai27**
## Description
I have implemented a rudimentary TopN Pushdown for the Mongo connector.
Pushdown is supported when all the SortItems have a SortOrder of either ASC_NULLS_FIRST or DESC_NULLS_LAST.
This is natively supported by MongoDB and also the Java Mongo Driver.
I have added comments in the code explaining the above limitations, and the rationale behind it and also pointing to the relevant sections in the MongoDB docs.
I have also added tests for the same.

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# MongoDB
* TBD. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by ebyhr — COMMENTED** [review]
Please squash commits into one and follow commit message guideline. 
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#commits-and-pull-requests

Also, please fix CI failure:
```
Error:    TestMongoConnectorTest.testSortItemsReflectedInExplain:145->AbstractTestQueryFramework.assertExplain:530->AbstractTestQueryFramework.assertExplain:535->AbstractTestQueryFramework.assertExplainAnalyze:563 
Expecting actual:
  "Trino version: testversion
Fragment 0 [SOURCE]
    Output layout: [name]
    Output partitioning: SINGLE []
    Output[columnNames = [name]]
    │   Layout: [name:varchar(25)]
    │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
    └─ TableScan[table = mongodb:MongoTableHandle[schemaTableName=tpch.nation, remoteTableName=RemoteTableName[databaseName=tpch, collectionName=nation], filter=Optional.empty, constraint=ALL, projectedColumns=[name:varchar(25)], limit=OptionalInt[5], sortItems=[SortItem[name='nationkey', sortOrder=DESC NULLS LAST]]]]
           Layout: [name:varchar(25)]
           Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           name := name:varchar(25)
```

**Comment by NikhilDesai27** [comment]
Hi @ebyhr,
Thanks for sharing your valuable feedback. I understand that raising a PR with just one commit is desirable, however, I did pull merges that happened in the meanwhile hoping it would make it easier for you all.
I noticed the CI issue and will fix it.
Please advise on how to squash the commits.
Thanks,
Nikhil.

**Review by Praveen2112 — COMMENTED** [review]
Thanks for working on this.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMetadata.java [hunk 1]
```diff
@@ -617,11 +622,99 @@ public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(Connect
...
+        }
+
+        log.info("MongoDB: applyTopN with the following sortItems");
+
+        // MongoDB allows a column/field to take a value of any type across documents.
+        // Hence, it's sort order factors in this type information as well.
+        // null is valued less than any other type and so on
+        // see, https://www.mongodb.com/docs/manual/reference/method/cursor.sort/#ascending-descending-sort
+        // For now, we only support ASC_NULLS_FIRST and DESC_NULLS_LAST, as these are implemented in MongoDB by default
+        for (SortItem si : sortItems) {
```
[the remark this answers is not in the document]
Comment (NikhilDesai27) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMetadata.java:667]: Hi @Praveen2112 ,
Thanks for your valuable feedback. I will incorporate them.
Thanks,
Nikhil Desai.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMetadata.java [hunk 2]
```diff
@@ -617,11 +622,99 @@ public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(Connect
...
+            return Optional.empty();
+        }
+
+        // MongoDB supports a maximum of 32 fields in it's cursor.sort() method (backed by $sort operator)
+        // see, https://www.mongodb.com/docs/manual/reference/method/cursor.sort/#limits
+        if (sortItems.size() > 32) {
+            return Optional.empty();
+        }
+
+        log.info("MongoDB: applyTopN with the following sortItems");
```
[the remark this answers is not in the document]
Comment (NikhilDesai27) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMetadata.java:660]: yes

**Comment by NikhilDesai27** [comment]
Hi @ebyhr and @Praveen2112,
I understand the reason why one of the CI tests is failing.
With TopN pushdown implemented, the `TopNPartial` step would no longer be present in the `EXPLAIN` output which is what the test deems as expected.
The test invokes a function from `AbstractTestQueryFramework`(assertExplainAnalyze) that only has functionality to check if a regex pattern is present in a string.
assertJ has a complementary functionality (doesNotContainPattern) too.
I can augment the `AbstractTestQueryFramework` or maybe there's an easy workaround like using negations in the pattern being passed in.
Please advise on how to proceed.

**Comment by NikhilDesai27** [comment]
Hi @Praveen2112,
You're right! I was planning to add three more test cases, totalling four, one for each `SortOrder`.
**Objective:** We then need to assert that TopN pushdown is happening in two of those test cases(ASC NULLS FIRST and DESC NULLS LAST) but not in the rest.
**Plan:** Introduce a new helper function `assertExplainDoesNotContain` that asserts a string doesn't contain a collection of regex patterns or sub-strings, to test that TopNPartial is not present when TopN pushdown happens. This will internally use assertJ's `doesNotContain`.

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 3]
```diff
@@ -68,7 +68,16 @@ support:
 * - `s3.http-proxy`
   - URL of a HTTP proxy server to use for connecting to S3.
 * - `s3.http-proxy.secure`
-  - Set to `true` to enable HTTPS for the proxy server..
+  - Set to `true` to enable HTTPS for the proxy server.
```
Comment (ebyhr) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:71]: Please revert unrelated changes.

Comment (NikhilDesai27) [reply docs/src/main/sphinx/object-storage/file-system-s3.md:71]: Hi @ebyhr,
Sorry, it's a result of syncing up my branch with the master from upstream.
Please advise on how to proceed. I am actually considering opening a new PR as my git tree is messy.
Thanks.

**Comment by mosabua** [comment]
@cla-bot check

**Comment by NikhilDesai27** [comment]
Hi @mosabua, I will open a new PR to make it easy for you all. I am on it and thanks for offering to help.
